### PR TITLE
Refactor IPv6 header length handling

### PR DESCRIPTION
**Pull request by ozhuraki**
Change the length to uint16_t and work with it through standard htons/ntohs() macros.

### CODE REVIEW

File: tests/net/6lo/src/main.c [hunk 1]
```diff
@@ -397,7 +396,6 @@ static struct net_6lo_data test_data_1 = {
 	.ipv6.vtc = 0x60,
 	.ipv6.tcflow = 0x00,
 	.ipv6.flow = 0x00,
-	.ipv6.len = { 0x00, 0x00 },
```
Comment (jukkar) [inline tests/net/6lo/src/main.c]: Even if it makes not difference here (as the variable is static), for consistency please set .len = 0 here like you did for couple of these testing variables below.

**Review by jukkar — CHANGES_REQUESTED** [review]
LGTM, one minor tweak could be done.

File: subsys/net/ip/ipv6.c [hunk 2]
```diff
@@ -2123,11 +2118,12 @@ int net_ipv6_send_ns(struct net_if *iface,
 	net_icmpv6_set_ns_hdr(pkt, ns_hdr);
 
 	if (is_my_address) {
+		u16_t *len = &NET_IPV6_HDR(pkt)->len;
```
Comment (pfalcon) [inline subsys/net/ip/ipv6.c]: I'd say, this gets hairy and hardly an improvement. Please use standard var arithmetics, without pointers.

File: tests/net/6lo/src/main.c [hunk 3]
```diff
@@ -435,7 +432,7 @@ static struct net_6lo_data test_data_3 = {
 	.ipv6.vtc = 0x60,
 	.ipv6.tcflow = 0x21,
 	.ipv6.flow = 0x3412,
-	.ipv6.len = { 0x00, 0x00 },
+	.ipv6.len = 0,
```
Comment (pfalcon) [inline tests/net/6lo/src/main.c]: Nitpick, but did you consider using htons(0) to avoid any concerns?

File: tests/net/6lo/src/main.c (hunk 1 above)
Comment (pfalcon) [inline tests/net/6lo/src/main.c]: Why here you remove this line completely, while bellow replace with ".ipv6.len = 0,".

File: tests/net/6lo/src/main.c [hunk 4]
```diff
@@ -582,7 +572,7 @@ static struct net_6lo_data test_data_11 = {
 	.ipv6.vtc = 0x61,
 	.ipv6.tcflow = 0x20,
 	.ipv6.flow = 0x00,
-	.ipv6.len = { 0x00, 0x00 },
+	.ipv6.len = 0,
```
Comment (pfalcon) [inline tests/net/6lo/src/main.c]: Same comments as above.

**Review by pfalcon — CHANGES_REQUESTED** [review]
Thanks for going thru these, but I'd say some cleanup needed.

File: tests/net/6lo/src/main.c (hunk 1 above)
Comment (ozhuraki) [reply tests/net/6lo/src/main.c]: OK, corrected.

File: tests/net/6lo/src/main.c (hunk 4 above)
Comment (ozhuraki) [reply tests/net/6lo/src/main.c]: OK, corrected.

File: subsys/net/ip/ipv6.c (hunk 2 above)
Comment (ozhuraki) [reply subsys/net/ip/ipv6.c]: As for my taste, the original snippet is slightly overloaded visually and using pointer here makes it easier to parse what's going on. But, don't really want on to argue on such matters of taste, OK, corrected.

File: tests/net/6lo/src/main.c [hunk 5]
```diff
@@ -397,7 +396,7 @@ static struct net_6lo_data test_data_1 = {
 	.ipv6.vtc = 0x60,
 	.ipv6.tcflow = 0x00,
 	.ipv6.flow = 0x00,
-	.ipv6.len = { 0x00, 0x00 },
+	.ipv6.len = htons(0),
```
Comment (jukkar) [inline tests/net/6lo/src/main.c]: IMHO this is totally useless, 0 is always 0 in every byte order. Well, we can go with this too as this makes no difference here.

**Review by jukkar — APPROVED** [review]

**Review by finikorg — APPROVED** [review]
LGTM

File: subsys/net/ip/ipv6.c [hunk 6]
```diff
@@ -2126,8 +2121,9 @@ int net_ipv6_send_ns(struct net_if *iface,
 		/* DAD */
 		net_ipaddr_copy(&NET_IPV6_HDR(pkt)->src,
 				net_ipv6_unspecified_address());
-
-		NET_IPV6_HDR(pkt)->len[1] -= llao_len;
+		NET_IPV6_HDR(pkt)->len = ntohs(NET_IPV6_HDR(pkt)->len) -
```
[the remark this answers is not in the document]
Comment (jukkar) [reply subsys/net/ip/ipv6.c]: > ntohs(NET_IPV6_HDR(pkt)->len) = htons(len);

What is this suppose to do, why is there ntohs() on left side?

[the remark this answers is not in the document]
Comment (pfalcon) [reply subsys/net/ip/ipv6.c]: Fixed copy-paste error. It's supposed to do a typical read-modify-write pattern.

[the remark this answers is not in the document]
Comment (pfalcon) [reply subsys/net/ip/ipv6.c]: > Sorry, my bad, just parsed it wrong. Corrected. Good catch.

Still not there, sorry. You now have:

~~~
NET_IPV6_HDR(pkt)->len = htons(NET_IPV6_HDR(pkt)->len -
						llao_len);
~~~

But you should take a ->len which is in networking order, convert to host order (you don't), subtract, convert back to netorder, store to ->. The best way to do it is apparently using an intermediate var, like my snippet in previous comment shows. Thanks.

[the remark this answers is not in the document]
Comment (ozhuraki) [reply subsys/net/ip/ipv6.c]: My bad. Missed conversion in setup_headers(). Corrected. Take another look, please.

**Review by pfalcon — APPROVED** [review]
Thanks!

**Pull request merged**